### PR TITLE
Add stddef.h

### PR DESCRIPTION
**Pull request by jyn514**
Closes https://github.com/jyn514/saltwater/issues/388. Gives a reasonable way to add additional header files.

Saltwater can, for the first time after 14 months, compile hello world with a GNU userland!

r? @pythondude325

### CODE REVIEW

File: src/analyze/expr.rs [hunk 1]
```diff
@@ -911,12 +911,15 @@ pub(super) fn literal(literal: Literal, location: Location) -> Expr {
...
+        expr.ctype.is_void_pointer() || expr.ctype.is_char_pointer() || expr.is_null()
+    };
     if left.ctype == right.ctype {
         true
-    } else if left.ctype.is_void_pointer() || left.ctype.is_char_pointer() || left.is_null() {
+    } else if is_convertible_to_any_pointer(left) && right.ctype.is_pointer() {
         left.ctype = right.ctype.clone();
         true
-    } else if right.ctype.is_void_pointer() || right.ctype.is_char_pointer() || right.is_null() {
+    } else if is_convertible_to_any_pointer(right) && left.ctype.is_pointer() {
```
Comment (pythondude325) [inline src/analyze/expr.rs:922]: I don't know what these do. Can I get a link to the relevant spec section?

File: src/lex/cpp.rs [hunk 2]
```diff
@@ -1066,15 +1066,33 @@ impl<'a> PreProcessor<'a> {
...
+                    .into();
+                (path, src)
+            }
+            Err(not_found) => {
+                let filename = match filename.file_name().and_then(|f| f.to_str()) {
+                    None => return Err(not_found),
+                    Some(f) => f,
+                };
+                match get_builtin_header(filename) {
+                    Some(file) => {
```
Comment (pythondude325) [inline src/lex/cpp.rs:1085]: Can you really override stddef.h? If so that's disgusting.

Comment (jyn514) [reply src/lex/cpp.rs:1085]: My reasoning here is that if the standard library provides stddef.h, that's probably more accurate than the one I hacked together. The standard doesn't treat the compiler and library separately so it's up to us which we want to do.

File: src/analyze/expr.rs (hunk 1 above)
Comment (jyn514) [reply src/analyze/expr.rs:922]: http://port70.net/~nsz/c/c11/n1570.html#6.5.15

Comment (jyn514) [reply src/analyze/expr.rs:923]: I also added a comment at the start of the function.

Comment (hdamron17) [reply src/analyze/expr.rs:923]: I feel like this should have been a separate pull request but oh well.

Comment (jyn514) [reply src/analyze/expr.rs:923]: Originally I had `#define va_arg(ap, type) (0 ? "va_arg is not currently supported by saltwater" : 0.0)`. But it turns out that I don't give _any_ sort of reasonable errors from macro expansion, so it didn't really matter.

```
stdarg_example.c:4:5 error: invalid program: incompatible types in ternary expression: 'char *' cannot be converted to 'double'
    va_start(args, n);
    ^^^^^^^^
```

Comment (jyn514) [reply src/analyze/expr.rs:923]: I improved the error messages a bit in the latest commit:

```c
stddef_example.c:5:21 error: invalid syntax: expected ')', got 'not_currently_supported'
    printf("%ul\n", offsetof(struct s { int i; }, i));
                    ^^^^^^^^
```

Still not great but better than before.

**Pull request merged**